src/bin/ion/commands/beta/generate/utils.rs

```diff
@@ -0,0 +1,158 @@
+use convert_case::{Case, Casing};
+use ion_schema::result::IonSchemaError;
+use serde::Serialize;
+use std::fmt::{Display, Formatter};
+use tera::Tera;
+use thiserror::Error;
+
+/// Represents code generation result
+pub type CodeGenResult<T> = Result<T, CodeGenError>;
+
+/// Represents an error found during code generation
+#[derive(Debug, Error)]
+pub enum CodeGenError {
+    #[error("{source:?}")]
+    IonError {
+        #[from]
+        source: IonSchemaError,
+    },
+    #[error("{source:?}")]
+    IoError {
+        #[from]
+        source: std::io::Error,
+    },
+    #[error("{source:?}")]
+    TeraError {
+        #[from]
+        source: tera::Error,
+    },
+    #[error("{description}")]
+    InvalidDataModel { description: String },
+}
+
+/// A convenience method for creating an CodeGen containing an CodeGenError::InvalidDataModel
+/// with the provided description text.
+pub fn invalid_data_model_error<T, S: AsRef<str>>(description: S) -> CodeGenResult<T> {
+    Err(CodeGenError::InvalidDataModel {
+        description: description.as_ref().to_string(),
+    })
+}
+
+/// Represents a field that will be added to generated data model.
+/// This will be used by the template engine to fill properties of a struct/classs.
+#[derive(Serialize)]
+pub struct Field {
+    pub(crate) name: String,
+    pub(crate) value: String,
+}
+
+/// Represents an import statement in a module file.
+/// this will be used by template engine to fill import statements of a type.
+#[derive(Serialize)]
+pub struct Import {
+    pub(crate) module_name: String,
+    pub(crate) type_name: String,
+}
+
+/// Represent the programming language for code generation.
+#[derive(Debug, Clone, PartialEq)]
+pub enum Language {
+    Rust,
+    Java,
+}
+
+impl Language {
+    pub fn file_extension(&self) -> &str {
+        match self {
+            Language::Rust => "rs",
+            Language::Java => "java",
+        }
+    }
+
+    pub fn file_name(&self, name: &str) -> String {
+        match self {
+            Language::Rust => name.to_case(Case::Snake),
+            Language::Java => name.to_case(Case::UpperCamel),
+        }
+    }
+}
+
+impl From<&str> for Language {
+    fn from(value: &str) -> Self {
+        match value {
+            "java" => Language::Java,
+            "rust" => Language::Rust,
+            _ => unreachable!("Unsupported programming language: {}, this tool only supports Java and Rust code generation.", value)
+        }
+    }
+}
+
+impl Display for Language {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Language::Rust => {
+                write!(f, "rust")
+            }
+            Language::Java => {
+                write!(f, "java")
+            }
+        }
+    }
+}
+
+/// Represents a context that will be used for code generation
+pub struct CodeGenContext {
```

Comment:
🗺️ All the necessary context for code generation is stored here. This will be used by the `generate_*()` methods in generate subcommand. It includes a `data_model` property that points to the data model determined based on constraints and is used to get appropriate template for the ISL type definition.